### PR TITLE
[SPARK-31009][SQL] Support json_object_keys function

### PR DESCRIPTION
**Pull request by iRakson**
### What changes were proposed in this pull request?
A new function `json_object_keys` is proposed in this PR. This function will return all the keys of the outmost json object. It takes Json Object as an argument.

- If invalid json expression is given, `NULL` will be returned.
- If an empty string or json array is given, `NULL` will be returned.
- If valid json object is given, all the keys of the outmost object will be returned as an array.
- For empty json object, empty array is returned.

We can also get JSON object keys using `map_keys+from_json`.  But `json_object_keys` is more efficient.
```
Performance result for json_object = {"a":[1,2,3,4,5], "b":[2,4,5,12333321]}

Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
JSON functions:                           Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
json_object_keys                                  11666          12361         673          0.9        1166.6       1.0X
from_json+map_keys                                15309          15973         701          0.7        1530.9       0.8X

```

### Why are the changes needed?
This function will help naive users in directly extracting the keys from json string and its fairly intuitive as well. Also its extends the functionality of spark-sql for json strings.

Some of the most popular DBMSs supports this function.
- PostgreSQL
- MySQL
- MariaDB

### Does this PR introduce any user-facing change?
Yes. Now users can extract keys of json objects using this function.

### How was this patch tested?
UTs added.

### CODE REVIEW

**Comment by iRakson** [comment]
@HyukjinKwon @cloud-fan @maropu @dongjoon-hyun

**Comment by MaxGekk** [comment]
@iRakson Could you show the use case of the function.

**Review by MaxGekk — COMMENTED** [review]
@iRakson Could you show the use case of the function.

**Comment by MaxGekk** [comment]
We can get the same result using existing functions:
```scala
scala> val df = Seq("""{"a":1, "b":2}""").toDF("json")
df: org.apache.spark.sql.DataFrame = [json: string]

scala> df.select(map_keys(from_json($"json", MapType(StringType, StringType)))).show
+-----------------+
|map_keys(entries)|
+-----------------+
|           [a, b]|
+-----------------+
```
What's the benefit of having the function in public API?

**Comment by iRakson** [comment]
> What's the benefit of having the function in public API?

- `json_object_keys` gives better performance than `map_keys+from_json`.
- Users are more familiar with `json_object_keys` as it supported by some of popular DBMSs, so they will find it easier to use.

**Comment by iRakson** [comment]
@HyukjinKwon @cloud-fan @maropu 
Please take a look at this as well.

**Comment by dongjoon-hyun** [comment]
ok to test

**Comment by dongjoon-hyun** [comment]
Do you know how much faster it is?
> json_object_keys gives better performance than map_keys+from_json.

**Comment by iRakson** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
@MaxGekk does it look good to you?

**Comment by maropu** [comment]
> Do you know how much faster it is?
>> json_object_keys gives better performance than map_keys+from_json.

Has the @dongjoon-hyun comment above been already adderssed? If this func has performance benefits, I think its better to show the perf. numbers in the PR description.

**Comment by iRakson** [comment]
> > Do you know how much faster it is?
> > > json_object_keys gives better performance than map_keys+from_json.
> 
> Has the @dongjoon-hyun comment above been already adderssed? If this func has performance benefits, I think its better to show the perf. numbers in the PR description.

I have updated the time required for `map_keys+from_json` and `json_object_keys`. I will also try to add benchmark for all the new JSON functions.
@maropu @HyukjinKwon

**Comment by iRakson** [comment]
> @MaxGekk does it look good to you?

@MaxGekk Kindly Review this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 1]
```diff
@@ -781,3 +783,64 @@ case class SchemaOfJson(
...
+    } catch {
+      case _: JsonProcessingException => null
+    }
+  }
+
+  private def getJsonKeys(parser: JsonParser, input: InternalRow): Any = {
+    var arrayBufferOfKeys = ArrayBuffer.empty[UTF8String]
+    // this handles `NULL` case
+    if (parser.nextToken() == null) {
+      throw new AnalysisException(s"$prettyName expect a JSON object but nothing is provided.")
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: Is the exception thrown at analysis phase? I think `AnalysisException` should be replaced by `RuntimeException`.

**Comment by MaxGekk** [comment]
The function works for literals but I got NPE on a column with JSON objects:
```scala
  test("json_object_keys") {
    spark.sql("""select json_object_keys('{"a": 1, "b": 2, "c": 3}')""").show(false)
    val df = Seq(
      """{"a":1}""",
      """{"a":1, "b":2}""",
      """{"a": 1, "b": 2, "c": 3}""").toDF("json")
    df.show(false)
    val dfKeys = df.selectExpr("json_object_keys(json)")
    dfKeys.show(false)
  }
```
```
+------------------------------------------+
|json_object_keys({"a": 1, "b": 2, "c": 3})|
+------------------------------------------+
|[a, b, c]                                 |
+------------------------------------------+

+------------------------+
|json                    |
+------------------------+
|{"a":1}                 |
|{"a":1, "b":2}          |
|{"a": 1, "b": 2, "c": 3}|
+------------------------+

06:28:11.089 ERROR org.apache.spark.executor.Executor: Exception in task 0.0 in stage 3.0 (TID 3)
java.lang.NullPointerException
	at org.apache.spark.sql.catalyst.InternalRow$.$anonfun$getAccessor$16(InternalRow.scala:152)
...
Caused by: java.lang.NullPointerException
	at org.apache.spark.sql.catalyst.InternalRow$.$anonfun$getAccessor$16(InternalRow.scala:152)
	at org.apache.spark.sql.catalyst.InternalRow$.$anonfun$getAccessor$16$adapted(InternalRow.scala:151)
	at org.apache.spark.sql.catalyst.expressions.BoundReference.eval(BoundAttribute.scala:41)
	at org.apache.spark.sql.catalyst.expressions.JsonObjectKeys.json$lzycompute(jsonExpressions.scala:812)
	at org.apache.spark.sql.catalyst.expressions.JsonObjectKeys.json(jsonExpressions.scala:812)
```

**Comment by iRakson** [comment]
> The function works for literals but I got NPE on a column with JSON objects:
> 
> ```scala
>   test("json_object_keys") {
>     spark.sql("""select json_object_keys('{"a": 1, "b": 2, "c": 3}')""").show(false)
>     val df = Seq(
>       """{"a":1}""",
>       """{"a":1, "b":2}""",
>       """{"a": 1, "b": 2, "c": 3}""").toDF("json")
>     df.show(false)
>     val dfKeys = df.selectExpr("json_object_keys(json)")
>     dfKeys.show(false)
>   }
> ```
> 
> ```
> +------------------------------------------+
> |json_object_keys({"a": 1, "b": 2, "c": 3})|
> +------------------------------------------+
> |[a, b, c]                                 |
> +------------------------------------------+
> 
> +------------------------+
> |json                    |
> +------------------------+
> |{"a":1}                 |
> |{"a":1, "b":2}          |
> |{"a": 1, "b": 2, "c": 3}|
> +------------------------+
> 
> 06:28:11.089 ERROR org.apache.spark.executor.Executor: Exception in task 0.0 in stage 3.0 (TID 3)
> java.lang.NullPointerException
> 	at org.apache.spark.sql.catalyst.InternalRow$.$anonfun$getAccessor$16(InternalRow.scala:152)
> ...
> Caused by: java.lang.NullPointerException
> 	at org.apache.spark.sql.catalyst.InternalRow$.$anonfun$getAccessor$16(InternalRow.scala:152)
> 	at org.apache.spark.sql.catalyst.InternalRow$.$anonfun$getAccessor$16$adapted(InternalRow.scala:151)
> 	at org.apache.spark.sql.catalyst.expressions.BoundReference.eval(BoundAttribute.scala:41)
> 	at org.apache.spark.sql.catalyst.expressions.JsonObjectKeys.json$lzycompute(jsonExpressions.scala:812)
> 	at org.apache.spark.sql.catalyst.expressions.JsonObjectKeys.json(jsonExpressions.scala:812)
> ```

I fixed this issue and also added a new test case for this particular issue.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 2]
```diff
@@ -796,3 +797,63 @@ case class SchemaOfJson(
...
+  """,
+  since = "3.1.0")
+case class JsonObjectKeys(child: Expression) extends UnaryExpression with CodegenFallback {
+  override def dataType: DataType = ArrayType(StringType)
+  override def nullable: Boolean = true
+  override def prettyName: String = "json_object_keys"
+
+  override def eval(input: InternalRow): Any = {
+    try {
+      lazy val json = child.eval(input).asInstanceOf[UTF8String]
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: why we need `lazy` here?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 3]
```diff
@@ -796,3 +797,63 @@ case class SchemaOfJson(
...
+
+/**
+ * A function which returns all the keys of outer JSON object.
+ */
+@ExpressionDescription(
+  usage = "_FUNC_(json_object) - returns all the keys of outer JSON object.",
+  arguments = """
+    Arguments:
+      * json_object - A JSON object is required as argument. `Null` is returned, if an invalid JSON
+          string is given. `Runtime Exception` is thrown, if null string or JSON array is given.
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: How about this?
```
      * json_object - A JSON object. If it is an invalid string, the function returns null.
              If it is a JSON array or null, a runtime exception will be thrown.
```
btw, the error handling is consistent with the other JSON functions?

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala [hunk 4]
```diff
@@ -710,4 +710,10 @@ class JsonFunctionsSuite extends QueryTest with SharedSparkSession {
       Seq(Row("string")))
   }
 
+  test("json_object_keys") {
+    val df = Seq("""{"a": 1, "b": 2, "c": 3}""".stripMargin)
+      .toDF("json")
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala]: nit: you don't need the line break here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala (hunk 3 above)
Comment (iRakson) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: I checked the other json functions. `IllegalArgumentException` is thrown by `from_json` for invalid inputs and it is better than `RuntimeException`.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by iRakson** [comment]
gentle ping @HyukjinKwon @maropu @MaxGekk

File: sql/core/src/test/resources/sql-tests/results/json-functions.sql.out [hunk 5]
```diff
@@ -346,6 +346,96 @@ org.apache.spark.sql.AnalysisException
...
+-- !query output
+NULL
+
+
+-- !query
+select json_object_keys('[1, 2, 3]')
+-- !query schema
+struct<>
+-- !query output
+java.lang.IllegalArgumentException
```
Comment (MaxGekk) [inline sql/core/src/test/resources/sql-tests/results/json-functions.sql.out]: Why do you throw `IllegalArgumentException` for `'[1, 2, 3]'` but NULL for `'{"key": 45, "random_string"}'`? It looks slightly inconsistent though both input are invalid.

**Review by MaxGekk — APPROVED** [review]
LGTM except one comment

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala [hunk 6]
```diff
@@ -710,4 +710,9 @@ class JsonFunctionsSuite extends QueryTest with SharedSparkSession {
       Seq(Row("string")))
   }
 
+  test("json_object_keys") {
+    val df = Seq("""{"a": 1, "b": 2, "c": 3}""".stripMargin).toDF("json")
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala]: Remove `stripMargin`

File: sql/core/src/test/resources/sql-tests/results/json-functions.sql.out (hunk 5 above)
Comment (iRakson) [reply sql/core/src/test/resources/sql-tests/results/json-functions.sql.out]: For invalid JSON strings NULL is thrown. `'[1, 2, 3]'` is a valid JSON string. But not a JSON object. So invalid argument.

Comment (iRakson) [reply sql/core/src/test/resources/sql-tests/results/json-functions.sql.out]: Regarding this issue. I am thinking of implementing a function `ISJSON()`. It will return `NULL` for `null`, `1` for valid JSON string and `0` for invalid JSON string.

This function is supported by most of the major DBMSs.

Currently we return `NULL` for invalid JSON.

In this PR as well as in #27759 , `NULL` is returned for `null` JSON strings. So behaviour might be confusing as we are returning `NULL` for invalid JSON strings as well.

**Review by HyukjinKwon — APPROVED** [review]
Will merge in few days if there are no more comments.

**Review by maropu — APPROVED** [review]
I have no more comment and it looks ok.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 7]
```diff
@@ -796,3 +797,64 @@ case class SchemaOfJson(
...
+
+  override def eval(input: InternalRow): Any = {
+    try {
+      val json = child.eval(input).asInstanceOf[UTF8String]
+      Utils.tryWithResource(CreateJacksonParser.utf8String(SharedFactory.jsonFactory, json)) {
+        parser => getJsonKeys(parser, input)
+      }
+    } catch {
+      case _: JsonProcessingException => null
+    }
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: Apparently, the current implementation seems to have the same limitations: `NullPointerException` and `ClassCastException`. Please add more unit test cases with `null` and `Integer` input.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/JsonExpressionsSuite.scala [hunk 8]
```diff
@@ -791,4 +791,42 @@ class JsonExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper with
...
+    val empty_json_object = """{}"""
+    val simple_json_object = """{"key": 1}"""
+    val another_simple_json_object = """{"key": "value", "key2": 2}"""
+    val json_object_with_array = """{"arrayKey": [1, 2, 3]}"""
+    val another_json_object_with_array = """{"key":[1,2,3,{"key":"value"},[1,2,3]]}"""
+    val complex_json_object = """{"f1":"abc","f2":{"f3":"a", "f4":"b"}}"""
+    val another_complex_json_object = """{"k1": [1, 2, {"key": 5}], "k2": {"key2": [1, 2]}}"""
+    val empty_json_array = """[]"""
+    val invalid_json_object = """{[1,2]}"""
+    val another_invalid_json_object = """{"key": 45, "random_string"}"""
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/JsonExpressionsSuite.scala]: Please use the simpler pattern which I commented as an example in your sister PR.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/JsonExpressionsSuite.scala [hunk 9]
```diff
@@ -791,4 +791,42 @@ class JsonExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper with
...
+    checkEvaluation(JsonObjectKeys(Literal(complex_json_object)), Seq("f1", "f2"))
+    checkEvaluation(JsonObjectKeys(Literal(another_complex_json_object)), Seq("k1", "k2"))
+    checkEvaluation(JsonObjectKeys(Literal(invalid_json_object)), null)
+    checkEvaluation(JsonObjectKeys(Literal(another_invalid_json_object)), null)
+
+    val exception = intercept[TestFailedException] {
+      checkEvaluation(JsonObjectKeys(Literal(null_object)), null)
+    }.getCause
+
+    assert(exception.isInstanceOf[IllegalArgumentException])
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/JsonExpressionsSuite.scala]: ditto. This should be checked at 818.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 10]
```diff
@@ -796,3 +797,64 @@ case class SchemaOfJson(
...
+
+/**
+ * A function which returns all the keys of outer JSON object.
+ */
+@ExpressionDescription(
+  usage = "_FUNC_(json_object) - returns all the keys of outer JSON object.",
+  arguments = """
+    Arguments:
+      * json_object - A JSON object. If it is an invalid string, the function returns null.
+          If it is a JSON array or null, a runtime exception will be thrown.
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: Is there a reason why you choose a runtime exception for `null`? In SQL world, for `null` input, `null` return is expected instead of a runtime exception.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala (hunk 7 above)
Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: I checked.
```
scala> sql("select json_object_keys(null)").show
java.lang.NullPointerException

scala> sql("select json_object_keys(1)").show
java.lang.ClassCastException: class java.lang.Integer cannot be cast to class org.apache.spark.unsafe.types.UTF8String (java.lang.Integer is in module java.base of loader 'bootstrap'; org.apache.spark.unsafe.types.UTF8String is in unnamed module of loader 'app')
```

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Hi, @iRakson . I'll review again after you revise this PR. Thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 11]
```diff
@@ -796,3 +797,64 @@ case class SchemaOfJson(
...
+    Examples:
+      > Select _FUNC_('{}');
+        []
+      > Select _FUNC_('{"key": "value"}');
+        ["key"]
+      > Select _FUNC_('{"f1":"abc","f2":{"f3":"a", "f4":"b"}}');
+        ["f1","f2"]
+  """,
+  since = "3.1.0")
+case class JsonObjectKeys(child: Expression) extends UnaryExpression with CodegenFallback {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:821]: please implement `ExpectsInputTypes` to specify the expected input types, so that the analyzer can do proper type check.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 12]
```diff
@@ -796,3 +797,64 @@ case class SchemaOfJson(
...
+  override def dataType: DataType = ArrayType(StringType)
+  override def nullable: Boolean = true
+  override def prettyName: String = "json_object_keys"
+
+  override def eval(input: InternalRow): Any = {
+    try {
+      val json = child.eval(input).asInstanceOf[UTF8String]
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:828]: most SQL functions return null for null input, I think this function should follow.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala (hunk 10 above)
Comment (iRakson) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:809]: Actually I used `null` here for null string(`''`).

**Comment by iRakson** [comment]
@dongjoon-hyun @cloud-fan 
I have tried to handle all the review comments. Kindly review the changes.

**Comment by dongjoon-hyun** [comment]
Thank you for updating, @iRakson .

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 13]
```diff
@@ -796,3 +797,71 @@ case class SchemaOfJson(
...
+
+/**
+ * A function which returns all the keys of outer JSON object.
+ */
+@ExpressionDescription(
+  usage = "_FUNC_(json_object) - returns all the keys of outer JSON object.",
+  arguments = """
+    Arguments:
+      * json_object - A JSON object. If it is an invalid string, the function returns null.
+          If it is a JSON array, an illegal argument exception will be thrown.
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:809]: Could you sync the PR description and this function accordingly? This function description looks too simple and the PR description seems wrong in the part `Analysis Exception`.
```
If invalid json expression is given, NULL will be returned.
If a null string or json array is given, an Analysis Exception is thrown.
If valid json object is given, all the keys of outer object will be returned as array.
For empty json object, empty array is returned.
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 14]
```diff
@@ -796,3 +797,71 @@ case class SchemaOfJson(
 
   override def prettyName: String = "schema_of_json"
 }
+
+/**
+ * A function which returns all the keys of outer JSON object.
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:802]: Since JSON can have a nested structure with multiple `inner`s and `outer`s, shall we use `the outmost` instead of `outer`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 15]
```diff
@@ -796,3 +797,71 @@ case class SchemaOfJson(
...
+        parser => getJsonKeys(parser, input)
+      }
+    } catch {
+      case _: JsonProcessingException => null
+    }
+  }
+
+  private def getJsonKeys(parser: JsonParser, input: InternalRow): Any = {
+    var arrayBufferOfKeys = ArrayBuffer.empty[UTF8String]
+    // this handles `NULL` case
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:846]: I understand that you meant `parser.nextToken() == null`, but shall we update the comment like `This handles empty string case`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 16]
```diff
@@ -796,3 +797,71 @@ case class SchemaOfJson(
...
+  with ExpectsInputTypes {
+
+  override def inputTypes: Seq[DataType] = Seq(StringType)
+  override def dataType: DataType = ArrayType(StringType)
+  override def nullable: Boolean = true
+  override def prettyName: String = "json_object_keys"
+
+  override def eval(input: InternalRow): Any = {
+    val json = child.eval(input).asInstanceOf[UTF8String]
+    if(json == null) {
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:831]: If you don't mind, please add a comment before this line.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 17]
```diff
@@ -796,3 +797,71 @@ case class SchemaOfJson(
...
+      case _: JsonProcessingException => null
+    }
+  }
+
+  private def getJsonKeys(parser: JsonParser, input: InternalRow): Any = {
+    var arrayBufferOfKeys = ArrayBuffer.empty[UTF8String]
+    // this handles `NULL` case
+    if (parser.nextToken() == null) {
+      throw new IllegalArgumentException(
+        s"$prettyName expect a JSON object but nothing is provided.")
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:849]: `expect` -> `expects`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala (hunk 13 above)
Comment (iRakson) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:809]: Actually PR description is not updated. I will update it. I will update the function description as well.

**Comment by iRakson** [comment]
@dongjoon-hyun 
I handled the comments. Please verify.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala (hunk 13 above)
Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 18]
```diff
@@ -796,3 +797,74 @@ case class SchemaOfJson(
...
+ * A function which returns all the keys of the outmost JSON object.
+ */
+@ExpressionDescription(
+  usage = "_FUNC_(json_object) - returns all the keys of the outmost JSON object as an array.",
+  arguments = """
+    Arguments:
+      * json_object - A JSON object. If a valid JSON object is given, all the keys of the outmost
+          object will be returned as an array. If it is an invalid JSON string, the function
+          returns null. If it is a JSON array or empty string, an illegal argument exception
+          will be thrown.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:811]: I'm confused. Why the behavior is different between invalid JSON and JSON array/empty string?

Comment (iRakson) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:811]: For invalid JSON we are returning `NULL` in all the other json functions. So for that i kept it like that.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:811]: Then we should return null for JSON array or empty string, as they are also invalid as the input of this function.

Comment (iRakson) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:811]: Ok, i will make the required changes here.
In that case, i guess i need to do the same in #27759 too ?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:811]: yea

**Comment by iRakson** [comment]
@cloud-fan I made the changes.
Kindly review the changes. I will also update the PR description accordingly.

**Comment by dongjoon-hyun** [comment]
Hi, @iRakson .
Could you resolve the conflicts?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/JsonExpressionsSuite.scala [hunk 19]
```diff
@@ -790,4 +790,24 @@ class JsonExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper with
...
+      ("""{"key": "value", "key2": 2}""", Seq("key", "key2")),
+      ("""{"arrayKey": [1, 2, 3]}""", Seq("arrayKey")),
+      ("""{"key":[1,2,3,{"key":"value"},[1,2,3]]}""", Seq("key")),
+      ("""{"f1":"abc","f2":{"f3":"a", "f4":"b"}}""", Seq("f1", "f2")),
+      ("""{"k1": [1, 2, {"key": 5}], "k2": {"key2": [1, 2]}}""", Seq("k1", "k2")),
+      ("{[1,2]}", null),
+      ("""{"key": 45, "random_string"}""", null),
+      ("", null),
+      ("[]", null),
+      ("""[{"key": "JSON"}]""", null)
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/JsonExpressionsSuite.scala:807]: Could you sort the test cases a little meaningfully, please?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/JsonExpressionsSuite.scala [hunk 20]
```diff
@@ -790,4 +790,24 @@ class JsonExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper with
...
+    ).foreach{
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/JsonExpressionsSuite.scala:808]: nit. `).foreach{` -> `).foreach {`

File: sql/core/src/test/resources/sql-tests/inputs/json-functions.sql [hunk 21]
```diff
@@ -58,5 +58,22 @@ select schema_of_json('{"c1":01, "c2":0.1}', map('allowNumericLeadingZeros', 'tr
...
 CREATE TEMPORARY VIEW jsonTable(jsonField, a) AS SELECT * FROM VALUES ('{"a": 1, "b": 2}', 'a');
 SELECT schema_of_json(jsonField) FROM jsonTable;
+
+-- json_object_keys
+select json_object_keys();
+select json_object_keys(null);
+select json_object_keys(200);
+select json_object_keys('');
+select json_object_keys('{"key": 1}');
+select json_object_keys('{}');
```
Comment (dongjoon-hyun) [inline sql/core/src/test/resources/sql-tests/inputs/json-functions.sql:68]: Shall we switch line 67 and 68?

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala [hunk 22]
```diff
@@ -710,4 +710,9 @@ class JsonFunctionsSuite extends QueryTest with SharedSparkSession {
       Seq(Row("string")))
   }
 
+  test("json_object_keys") {
+    val df = Seq("""{"a": 1, "b": 2, "c": 3}""").toDF("json")
+    val dfKeys = df.selectExpr("json_object_keys(json)")
+    checkAnswer(dfKeys, Row(Array("a", "b", "c")))
+  }
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala:717]: We can revert this change from this PR since we have a superset already.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 23]
```diff
@@ -796,3 +797,67 @@ case class SchemaOfJson(
 
   override def prettyName: String = "schema_of_json"
 }
+
+/**
+ * A function which returns all the keys of the outmost JSON object.
+ */
+@ExpressionDescription(
+  usage = "_FUNC_(json_object) - returns all the keys of the outmost JSON object as an array.",
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:805]: nit. `returns` -> `Returns`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 24]
```diff
@@ -796,3 +797,67 @@ case class SchemaOfJson(
...
+    if(json == null) {
+      return null
+    }
+
+    try {
+      Utils.tryWithResource(CreateJacksonParser.utf8String(SharedFactory.jsonFactory, json)) {
+        parser => getJsonKeys(parser, input)
+      }
+    } catch {
+      case _: JsonProcessingException => null
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:842]: Just a question: There is no need to handle `IOException` in this PR?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 25]
```diff
@@ -796,3 +797,67 @@ case class SchemaOfJson(
...
+    }
+  }
+
+  private def getJsonKeys(parser: JsonParser, input: InternalRow): Any = {
+    var arrayBufferOfKeys = ArrayBuffer.empty[UTF8String]
+    // return null if an empty string or any other valid JSON string is encountered
+    if (parser.nextToken() == null || parser.currentToken() != JsonToken.START_OBJECT) {
+      return null
+    }
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:851]: Shall we move this exception handling part (line 849 ~ 851) from this `getJsonKey` into line 839 (outside of this function)?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala (hunk 24 above)
Comment (iRakson) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:842]: Yes, `nextToken()` throws `IOException`.  I will update it with other changes.

**Comment by iRakson** [comment]
@dongjoon-hyun I resolved the conflicts and also have tried to handle all the review comments. 
Kindly review.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @iRakson and all.
Merged to master for Apache Spark 3.1.0.

**Comment by svmgarg** [comment]
> We can get the same result using existing functions:
> 
> ```scala
> scala> val df = Seq("""{"a":1, "b":2}""").toDF("json")
> df: org.apache.spark.sql.DataFrame = [json: string]
> 
> scala> df.select(map_keys(from_json($"json", MapType(StringType, StringType)))).show
> +-----------------+
> |map_keys(entries)|
> +-----------------+
> |           [a, b]|
> +-----------------+
> ```
> 
> What's the benefit of having the function in public API?

We don't need to define the schema in Json_object_keys
